source/common/network/lc_trie.h

```diff
@@ -146,27 +158,108 @@ class LcTrie {
-      if (nested_prefixes_ == nullptr) {
-        nested_prefixes_ = std::make_shared<std::vector<IpPrefix>>();
-      }
-      nested_prefixes_->push_back(other);
-    }
-
     // The address represented either in Ipv4(uint32_t) or Ipv6(asbl::uint128).
     IpType ip_{0};
     // Length of the cidr range.
-    int length_;
-    // TODO(ccaraman): Support more than one tag per entry.
-    // Tag for this entry.
-    std::string tag_;
-    // Other prefixes nested under this one. If an LC trie lookup matches on this
-    // prefix, the lookup will scan the nested prefixes to see if any of them match,
-    // too. This situation is rare, so to save memory in the common case the
-    // nested_prefixes field is a pointer to a vector rather than an inline vector.
-    // TODO(brian-pane) switch to a trie of nested prefixes, to ensure sublinear-time
-    // searching even in situations where there are a lot of nested prefixes.
-    std::shared_ptr<std::vector<IpPrefix>> nested_prefixes_;
+    uint32_t length_{0};
+    // Tag(s) for this entry.
+    TagSet tags_;
+  };
+
+  /**
+   * Binary trie used to simplify the construction of Level Compressed Tries.
+   * This data type supports two operations:
+   *   1. Add a prefix to the trie.
+   *   2. Push the prefixes the leaves of the trie.
+   * That second operation produces a new set of prefixes that yield the same
+   * match results as the original set of prefixes from which the BinaryTrie
+   * was constructed, but with an important difference: the new prefixes are
+   * guaranteed not to be nested within each other. That allows the use of the
+   * classic LC Trie construction algorithm, which is fast and (relatively)
+   * simple but does not work properly with nested prefixes.
+   */
+  template <class IpType, uint32_t address_size = CHAR_BIT * sizeof(IpType)> class BinaryTrie {
+  public:
+    BinaryTrie() : root_(std::make_unique<Node>()) {}
+
+    /**
+     * Add a CIDR prefix and associated tag to the binary trie. If an entry already
+     * exists for the prefix, merge the tag into the existing entry.
+     */
+    void insert(const IpPrefix<IpType>& prefix) {
+      Node* node = root_.get();
+      for (uint32_t i = 0; i < prefix.length_; i++) {
+        auto bit = static_cast<uint32_t>(extractBits(i, 1, prefix.ip_));
+        NodePtr& next_node = node->children[bit];
+        if (next_node == nullptr) {
+          next_node = std::make_unique<Node>();
+        }
+        node = next_node.get();
+      }
+      if (node->tags == nullptr) {
+        node->tags = std::make_shared<TagSet>();
+      }
+      node->tags->insert(prefix.tags_.begin(), prefix.tags_.end());
+    }
+
+    /**
+     * Update each node in the trie to inherit/override its ancestors' tags,
+     * and then push the prefixes in the binary trie to the leaves so that:
+     *  1) each leaf contains a prefix, and
+     *  2) no leaf's prefix is nested within another leaf's prefix.
```

Comment:
How can leafs nest within each other? I understand in the above example that internal nodes can nest with leafs, but comment here seems a bit harder to parse.

Reply:
Leaves cannot nest within other leaves. That's a fundamental property of trees, and it's why leaf-pushing is used in routing code.

Reply:
Yes, I realize what trees are @brian-pane. Can you just provide an example of what the world would look like if (2) were not true? It's unclear to me what you are saying here.

Reply:
I added new text for (2) that focuses less on the fact that leaves aren't nested under other leaves and more on how that property leads to a set of non-overlapping prefixes.